cmd/gcp-controller-manager/readonly_resource_csr_approver.go

```diff
@@ -0,0 +1,202 @@
+	capi "k8s.io/api/certificates/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cloud-provider-gcp/pkg/csrmetrics"
+	"k8s.io/klog/v2"
+	certutil "k8s.io/kubernetes/pkg/apis/certificates/v1"
+	"k8s.io/kubernetes/pkg/controller/certificates"
+)
+
+const (
+	ReadOnlyResourceAccess            = "TODO Define a name"
+	ReadOnlyResourceSignerName        = "gke.io/gke-kubelet-readonly-client"
+	ReadOnlyResourceCSRRequestMetrics = "readonly_resource_csr_request"
+)
+
+type readonlyResourceCSRRequestContent struct {
+	ctx    **controllerContext
+	csr    *capi.CertificateSigningRequest
+	x509cr *x509.CertificateRequest
+}
+
+type readonlyResourceCSRRequestResult struct {
+	result  bool
+	err     error
+	message string
+}
+
+type challengeFunc func(content readonlyResourceCSRRequestContent) readonlyResourceCSRRequestResult
+
+// -----------------------Approver-----------------------------
+// readonlyApprover handles approval/denial of CSRs access to read only resource
+// There is a list of approvers, requst must get all approval
+type readonlyResourceApprover struct {
+	ctx        *controllerContext
+	validators []readonlyResourceCsrValidator
+}
+
+func newReadonlyResourceApprover(ctx *controllerContext) readonlyResourceApprover {
+	return readonlyResourceApprover{
+		ctx:        ctx,
+		validators: newReadonlyResourceCsrValidator(),
+	}
+}
+
+func (approver *readonlyResourceApprover) handle(ctx context.Context, csr *capi.CertificateSigningRequest) error {
+	if csr == nil || csr.Spec.SignerName != ReadOnlyResourceSignerName {
+		return nil
+	}
+	recordMetric := csrmetrics.ApprovalStartRecorder(ReadOnlyResourceCSRRequestMetrics)
+	if len(csr.Status.Certificate) != 0 {
+		return nil
+	}
+	if approved, denied := certificates.GetCertApprovalCondition(&csr.Status); approved || denied {
+		return nil
+	}
+	klog.Infof("approver got CSR %q", csr.Name)
+
+	x509cr, err := certutil.ParseCSR(csr.Spec.Request)
+	if err != nil {
+		recordMetric(csrmetrics.ApprovalStatusParseError)
+		return fmt.Errorf("unable to parse csr %q: %v", csr.Name, err)
+	}
+
+	for _, validator := range approver.validators {
+		recordValidatorMetric := csrmetrics.ApprovalStartRecorder(validator.authFlowLabel)
+		content := readonlyResourceCSRRequestContent{
+			ctx:    &approver.ctx,
+			csr:    csr,
+			x509cr: x509cr,
+		}
+		challengeResult := validator.challenge(content)
+
+		switch {
+		case challengeResult.err != nil:
+			return fmt.Errorf("validating CSR %q: %v", csr.Name, err)
+		case challengeResult.result == false:
+			klog.Infof("validator %q: denied CSR %q", validator.name, csr.Name)
+			recordValidatorMetric(csrmetrics.ApprovalStatusDeny)
+			return approver.updateCSR(csr, false, challengeResult.message)
+		case challengeResult.result == true:
+			klog.Infof("validator %q: approved CSR %q", validator.name, csr.Name)
+			recordValidatorMetric(csrmetrics.ApprovalStatusApprove)
+		}
+		klog.Infof("CSR %q validation passed", csr.Name)
```

Comment:
log the validator function here as well, otherwise you'll get three identical messages for the same csr. 
also, consider logging this only on error, or at a higher verbosity.

Reply:
done